src/external/ACSAdapter/AMSFileScanner.ts

```diff
@@ -0,0 +1,138 @@
+        }
+    }
+
+    public retrieveFileScanResult(id: string): FileScanResult | undefined {
+        return this.scanResults?.get(id);
+    }
+
+    public addOrUpdateFile(id: string, fileMetadata: FileMetadata, scan: FileScanResponse): void {
+        const scanResult = this.retrieveFileScanResult(id);
+        this.scanResults?.set(id, {
+            ...(scanResult || {}),
+            fileMetadata,
+            scan
+        });
+    }
+
+    public addNext(id: string, next: Function): void {
+        const scanResult = this.scanResults?.get(id);
+        if (scanResult) {
+            this.scanResults?.set(id, {...scanResult, next});
+        }
+    }
+
+    public addActivity(id: string, activity: any): void {  // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+        const scanResult = this.scanResults?.get(id);
+        if (scanResult) {
+            this.scanResults?.set(id, {...scanResult, activity});
+        }
+    }
+
+    public async scanFileCallback(scanResult: FileScanResult, id: string): Promise<void> {
+        const {fileMetadata, next, activity} = scanResult;
+
+        if (scanResult?.scan?.status === AMSDownloadStatus.INPROGRESS) {
+            try {
+                const response: any = await this.amsClient.getViewStatus(fileMetadata); // eslint-disable-line @typescript-eslint/no-explicit-any
+                const {view_location, scan} = response;
+
+                this.addOrUpdateFile(id, scanResult.fileMetadata, scan);
+
+                if (scan.status === AMSDownloadStatus.PASSED && next && activity) {
+                    let blob: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+                    try {
+                        blob = await this.amsClient.getView(fileMetadata, view_location); // eslint-disable-line @typescript-eslint/no-explicit-any
+                    } catch (error) {
+                        console.error(error);
+                    }
+
+                    const file = new File([blob], fileMetadata.name as string, { type: fileMetadata.type});
+                    const attachmentData = await activityUtils.getAttachments([file]);
+                    const attachmentSizes = await activityUtils.getAttachmentSizes([file]);
+
+                    const index = activity.attachments.findIndex((attachment: any) => (attachment.name === fileMetadata.name)); // eslint-disable-line @typescript-eslint/no-explicit-any
+                    activity.channelData.fileScan[index] = scan;
+                    activity.attachments[index] = attachmentData[0];
+                    activity.channelData.attachmentSizes[index] = attachmentSizes[0];
+
+                    const hasMultipleAttachments = index > 0;
+                    if (hasMultipleAttachments) {
+                        const {metadata: {amsreferences}} = activity.channelData;
+                        const fileIds = JSON.parse(amsreferences);

```

Comment:
it seems at this point that elements are strongly coupled, but we have no validation in case the objects are null,

Reply:
The `activity` that's been retrieved is added by `createFileScanIngressMiddleware`. `createFileScanIngressMiddleware` has a `amsreferences` before adding it in the scanner.